src/representation_theory/monoid_algebra_basis.lean

```diff
@@ -0,0 +1,213 @@
+import representation_theory.basic
+
+/-!
+# The structure of the `k[G]`-module `k[Gⁿ]`
+
+This file contains API for the `k[G]`-module `k[Gⁿ]`, where `k` is a commutative ring, `G` is
+(mainly) a group, and the module structure is induced by the diagonal action of `G` on `Gⁿ.`
+
+In particular, we define `k[G]`-linear maps between `k[Gⁿ⁺¹]` and `k[G] ⊗ₖ k[Gⁿ]` which we will
+later show define an isomorphism. From this we will deduce that `k[Gⁿ⁺¹]` is free over `k[G]`.
+
+The freeness of these modules means we can use them to build a projective resolution of the
+(trivial) `k[G]`-module `k`, which is useful in group cohomology.
+
+## Main definitions
+
+  * mul_action_to_ρ
+  * mul_action_to_Rep
+  * to_tensor
+  * of_tensor
+
+## Implementation notes
+
+We bundle the type `k[Gⁿ]` - or more generally `k[H]` when `H` has `G`-action - and the
+`k[G]`-module structure together as a term of type `Rep k G`, and we call it
+`mul_action_to_Rep k G H.` The representation `ρ: G →* End(k[H])` is then induced by the `G`-action
+on `H`. This is so we only define a `k[G]`-module instance on `mul_action_to_Rep k G H` and not
+directly on `monoid_algebra k H`, which could cause diamonds.
+
+-/
+
+noncomputable theory
+universes v u
+variables {k G : Type u} [comm_ring k]
+
+open monoid_algebra
+namespace Rep
+
+section
+variables [monoid G]
+
+/- Not sure where to put the next several things; either `representation_theory.Rep` or
+`representation_theory.basic`, neither of which import the other. -/
+
+/-- A term of type `Rep k G` has the natural structure of a `k[G]`-module, with an element
+`g : G` acting by `g • x = (ρ g) x` -/
+instance module' (V : Rep k G) : module (monoid_algebra k G) V :=
+representation.as_module V.ρ
+
+variables {V : Type u} [add_comm_group V] [module k V] {ρ : representation k G V}
+
+lemma smul_def (g : monoid_algebra k G) (x : Rep.of ρ) : g • x = lift k G _ ρ g x := rfl
+
+@[simp] lemma single_smul (g : G) (r : k) (x : Rep.of ρ) :
+  ((•) : monoid_algebra k G → Rep.of ρ → Rep.of ρ)
+    (finsupp.single g r : monoid_algebra k G) x = r • ρ g x :=
+by rw [smul_def, lift_single]; refl
+
+instance : smul_comm_class (monoid_algebra k G) k (Rep.of ρ) := has_scalar.comp.smul_comm_class _
+
+instance : is_scalar_tower k (monoid_algebra k G) (Rep.of ρ) :=
+{ smul_assoc := λ x y z, by simp only [smul_def, alg_hom.map_smul, linear_map.smul_apply] }
+
+variables (k G) (H : Type u) [mul_action G H]
+
+/-- A `G`-action on `H` induces a representation `G →* End(k[H])` in the natural way. -/
+@[simps] def mul_action_to_ρ : representation k G (monoid_algebra k H) :=
+{ to_fun := λ g, finsupp.lmap_domain k k ((•) g),
+  map_one' := by { ext x y, dsimp, simp },
+  map_mul' := λ x y, by { ext z w, simp [mul_smul] }}
+
+/-- Given a `G`-action on `H`, this is `k[H]` bundled with the natural representation
+`G →* End(k[H])` as a term of type `Rep k G.` -/
+abbreviation mul_action_to_Rep : Rep k G := Rep.of (mul_action_to_ρ k G H)
```

Comment:
Maybe these could be in namespaces, as `representation.of_mul_action` and `Rep.of_mul_action`?